cmd/admin-monitor.go

```diff
@@ -0,0 +1,202 @@
+	MemUsage *madmin.ServerMemUsageInfo `json:"mem,omitempty"`
+}
+
+func (s serverMonitorMessage) JSON() string {
+	s.Monitor = "success"
+	statusJSONBytes, e := json.MarshalIndent(s, "", "  ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON")
+
+	return string(statusJSONBytes)
+}
+func (s serverMonitorMessage) String() string {
+	msg := ""
+
+	dot := "●"
+
+	// if offline
+	if s.Service == "off" {
+		msg += fmt.Sprintf("%s  %s\n", console.Colorize("MonitorFail", dot), s.Addr)
+		msg += fmt.Sprintf("   Server is %s", console.Colorize("MonitorFail", "offline"))
+		return msg
+	}
+
+	if s.Err != "" {
+		msg += fmt.Sprintf("%s  %s\n", console.Colorize("MonitorFail", dot), s.Addr)
+		msg += fmt.Sprintf("   Server is %s\n", console.Colorize("MonitorFail", "offline"))
+		msg += fmt.Sprintf("    Error : %s", s.Err)
+		return msg
+	}
+
+	// Print server title
+	msg += fmt.Sprintf("%s  %s    \n", console.Colorize("Monitor", dot), s.Addr)
+	msg += "\n"
+
+	msg += fmt.Sprintf("%s        min        avg      max\n", console.Colorize("Monitor", "   CPU"))
+	for i := range s.CPULoad.Load {
+		msg += fmt.Sprintf("   current    %.2f%%      %.2f%%    %.2f%%\n", s.CPULoad.Load[i].Min, s.CPULoad.Load[i].Avg, s.CPULoad.Load[i].Max)
+		msg += fmt.Sprintf("   historic   %.2f%%      %.2f%%    %.2f%%\n", s.CPULoad.HistoricLoad[i].Min, s.CPULoad.HistoricLoad[i].Avg, s.CPULoad.HistoricLoad[i].Max)
+		msg += "\n"
+	}
+
+	// Mem section
+	msg += fmt.Sprintf("%s        usage\n", console.Colorize("Monitor", "   MEM"))
+	for i := range s.MemUsage.Usage {
+		msg += fmt.Sprintf("   current    %s\n", humanize.IBytes(s.MemUsage.Usage[i].Mem))
+		msg += fmt.Sprintf("   historic   %s\n", humanize.IBytes(s.MemUsage.HistoricUsage[i].Mem))
+		msg += "\n"
+	}
+
+	return msg
+}
+
+func mainAdminMonitor(ctx *cli.Context) error {
+	checkAdminMonitorSyntax(ctx)
+
+	// set the console colors
+	console.SetColor("Monitor", color.New(color.FgGreen, color.Bold))
+	console.SetColor("MonitorDegraded", color.New(color.FgYellow, color.Bold))
+	console.SetColor("MonitorFail", color.New(color.FgRed, color.Bold))
+
+	// Get the alias
+	args := ctx.Args()
+	aliasedURL := args.Get(0)
+
+	// Create a new minio admin client
+	client, err := newAdminClient(aliasedURL)
+	fatalIf(err, "Cannot get a configured admin connection.")
+
+	printOfflineMessage := func() {
+		printMsg(serverMonitorMessage{
+			Addr:    aliasedURL,
+			Service: "off",
+		})
+	}
+
+	processErr := func(e error) error {
+		switch e.(type) {
+		case *json.SyntaxError:
+			printOfflineMessage()
+			return e
+		case *url.Error:
+			printOfflineMessage()
+			return e
+		default:
+			// If the error is not nil and unrecognized, just print it and exit
+			fatalIf(probe.NewError(e), "Cannot get service status.")
+		}
+		return nil
+	}
+	// Fetch info of all CPU loads (all minio server instances)
+	cpuMonitors, e := client.ServerCPULoadInfo()
+	if err := processErr(e); err != nil {
+		// exit immediately if error encountered
+		return nil
```

Comment:
Why are we returning `nil` here when we have actually encountered an error?

Reply:
I'm returning `nil` to keep up the aesthetics of our cli output. If I returned err, this is how the output looks.

```
ubuntu@ip-172-31-27-137:~/code/go/src/github.com/minio/mc$ ./mc  admin monitor local11111                                                                                            
●  local11111                                                                                                                                                                        
   Server is offline                                                                                                                                                                 
Get http://127.0.0.1:11111/minio/admin/v1/performance?perfType=cpu: dial tcp 127.0.0.1:11111: connect: connection refused 
```

Reply:
Fixed the formatting to look like this:

```
●  172.17.0.2:9000                                                                                                                                                                   
                                                                                                                                                                                     
   CPU        min        avg      max                                                                                                                                                
   current    0.08%      0.10%    0.10%                                                                                                                                              
   historic   0.02%      0.13%    4.36%                                                                                                                                              
                                                                                                                                                                                     
   MEM        usage                                                                                                                                                                  
   current    71MiB                                                                                                                                                                  
   historic   71MiB                                                                                                                                                                  
                                                                                                                                                                                     
                                                                                                                                                                                     
●  172.17.0.3:9000                                                                                                                                                                   
   Server is offline                                                                                                                                                                 
   Error : Post http://172.17.0.3:9000/minio/peer/remote: dial tcp 172.17.0.3:9000: connect: connection refused                                                                      
                                                                                                                                                                                     
●  172.17.0.4:9000                                                                                                                                                                   
                                                                                                                                                                                     
   CPU        min        avg      max                                                                                                                                                
   current    0.07%      0.08%    0.09%                                                                                                                                              
   historic   0.02%      0.13%    4.38%                                                                                                                                              
                                                                                                                                                                                     
   MEM        usage                                                                                                                                                                  
   current    71MiB                                                                                                                                                                  
   historic   71MiB                                                                                                                                                                  
                                                                                                                                                                                     
                                                                                                                                                                                     
●  172.17.0.5:9000                                                                                                                                                                   
   Server is offline                                                                                                                                                                 
   Error : Post http://172.17.0.5:9000/minio/peer/remote: dial tcp 172.17.0.5:9000: connect: no route to host         
```